### PR TITLE
service: am: Add support for LLE Software Keyboard Applet

### PR DESCRIPTION
**Pull request by dima-xd**
How it looks:
image

### CODE REVIEW

File: src/yuzu/main.cpp [hunk 1]
```diff
@@ -4341,6 +4342,29 @@ void GMainWindow::OnMiiEdit() {
     BootGame(filename);
 }
 
+void GMainWindow::OnSoftwareKeyboard() {
```
Comment (liamwhite) [inline src/yuzu/main.cpp:4345]: Is there any particular use for being able to boot this applet from the frontend? (other am changes are fine)

Comment (dima-xd) [reply src/yuzu/main.cpp:4345]: > Is there any particular use for being able to boot this applet from the frontend? (other am changes are fine)

I think no (maybe only to click buttons just for fun). So it's up to you. But for my opinion it's not needed at all

Comment (dima-xd) [reply src/yuzu/main.cpp]: removed

File: src/core/hle/service/am/am.cpp [hunk 2]
```diff
@@ -1759,6 +1779,57 @@ void ILibraryAppletSelfAccessor::PushInShowMiiEditData() {
...
+        .initial_string_length{},
+        .user_dictionary_offset{},
+        .user_dictionary_entries{},
+        .use_text_check = false,
+    };
+
+    Applets::SwkbdConfigNew swkbd_config_new{};
+
+    std::vector<u8> argument_data(sizeof(arguments));
+    std::vector<u8> swkbd_data(sizeof(swkbd_config));
```
Comment (german77) [inline src/core/hle/service/am/am.cpp:1820]: `swkbd_data` should be the size of `swkbd_config + swkbd_config_new`

File: src/core/hle/service/am/am.cpp [hunk 3]
```diff
@@ -1759,6 +1779,57 @@ void ILibraryAppletSelfAccessor::PushInShowMiiEditData() {
...
+    std::vector<u8> swkbd_new_data(sizeof(swkbd_config_new));
```
Comment (german77) [inline src/core/hle/service/am/am.cpp:1821]: This isn't new data is a work buffer. The size is `swkbd_config.initial_string_length * sizeof(char16_t))` and it should contain the initial text string.

File: src/core/hle/service/am/am.cpp [hunk 4]
```diff
@@ -1759,6 +1779,57 @@ void ILibraryAppletSelfAccessor::PushInShowMiiEditData() {
...
+
+    std::memcpy(argument_data.data(), &arguments, sizeof(arguments));
+    std::memcpy(swkbd_data.data(), &swkbd_config, sizeof(swkbd_config));
+    std::memcpy(swkbd_new_data.data(), &swkbd_config_new, sizeof(Applets::SwkbdConfigNew));
```
Comment (german77) [inline src/core/hle/service/am/am.cpp:1825]:
```suggestion
    std::memcpy(swkbd_data.data() +sizeof(swkbd_config) , &swkbd_config_new, sizeof(Applets::SwkbdConfigNew));
```

File: src/core/hle/service/am/am.cpp [hunk 5]
```diff
@@ -1759,6 +1779,57 @@ void ILibraryAppletSelfAccessor::PushInShowMiiEditData() {
...
+    std::memcpy(&swkbd_config_new, swkbd_data.data() + sizeof(Applets::SwkbdConfigCommon),
```
Comment (german77) [inline src/core/hle/service/am/am.cpp:1826]: Again this should be a work buffer
```suggestion
    std::memcpy(&swkbd_work_buffer, &initial_string, swkbd_config.initial_string_length * sizeof(char16_t));
```

File: src/core/hle/service/am/am.cpp [hunk 6]
```diff
@@ -1759,6 +1779,57 @@ void ILibraryAppletSelfAccessor::PushInShowMiiEditData() {
...
+                sizeof(Applets::SwkbdConfigNew));
+    queue_data.emplace_back(std::move(argument_data));
+    queue_data.emplace_back(std::move(swkbd_data));
+    queue_data.emplace_back(std::move(swkbd_new_data));
```
Comment (german77) [inline src/core/hle/service/am/am.cpp:1830]: ditto work buffer

**Review by liamwhite — APPROVED** [review]

**Pull request merged**